File: inspector/bif.cpp
#include "bif.h"
#include "mainwindow.h"
#include <zlib.h>

BIF::BIF(QString path) {
  handle = QSharedPointer<Handle>(new Handle(path));
  if (!handle->exists())
    throw FileError(QString("Can't open %1").arg(path));
}


QSharedPointer<Handle> BIF::get(quint16 tileset, quint16 index) {
  handle->seek(0);
  QString key = handle->r4();
  if (key != "BIFF" && key != "BIFC")
    throw FileError("Not a BIF file");
  bool compressed = key == "BIFC";

  QString ver = handle->r4();
  if (ver != "V1  " && ver != "V1.0")
    throw FileError("Invalid BIF version");

  if (!tileset) {
    if (compressed) {
      auto h = getBlock(16, 4);
      quint32 fileOffs = h->r32() + index * 16;
      h = getBlock(fileOffs, 16);
      h->skip(4); //locator
      quint32 offset = h->r32();
      quint32 size = h->r32();
      return getBlock(offset, size);
    } else {
      handle->skip(8); //num files and num tiles
      quint32 fileOffs = handle->r32() + index * 16;
      handle->seek(fileOffs);
      handle->skip(4); //locator
      quint32 offset = handle->r32();
      quint32 size = handle->r32();
      return QSharedPointer<Handle>(new Handle(handle, offset, size));
    }
  } else {
    if (compressed) {
      auto h = getBlock(8, 4);
      quint32 numFiles = h->r32();
      h = getBlock(16, 4);
      quint32 fileOffs = h->r32() + numFiles * 16 + tileset * 20;
      h = getBlock(fileOffs, 20);
      h->skip(4); //locator
      quint32 offset = h->r32();
      quint32 numTiles = h->r32();
      quint32 size = h->r32();
      return getBlock(offset, size * numTiles);
    } else {
      quint32 numFiles = handle->r32();
      handle->skip(4); //num tiles
      quint32 fileOffs = handle->r32() + numFiles * 16 + tileset * 20;
      handle->seek(fileOffs);
      handle->skip(4); //locator
      quint32 offset = handle->r32();
      quint32 numTiles = handle->r32();
      quint32 size = handle->r32();
      return QSharedPointer<Handle>(new Handle(handle, offset, size * numTiles));
    }
  }
}

QSharedPointer<Handle> BIF::getBlock(qint64 offset, quint64 length) {
  handle->seek(12);
  quint32 cur = 0;
  quint32 cLen = 0;
  ulong uLen = 0;
  while (cur <= offset) {
    handle->skip(cLen);
    uLen = handle->r32();
    cLen = handle->r32();
    cur += uLen;
  }

  QByteArray arr;
  quint32 delta = offset - (cur - uLen);
  quint32 dLen = 0;
  while (dLen < length) {
    Bytef *in = (Bytef *)handle->read(cLen);
    Bytef *out = new Bytef[uLen];
    uncompress(out, &uLen, in, cLen);
    uLen -= delta;
    if (uLen + dLen > length)
      uLen = length - dLen;
    arr.append((const char *)out + delta, uLen);
    dLen += uLen;
    delta = 0;
    delete [] out;
    uLen = handle->r32();
    cLen = handle->r32();
  }
  return QSharedPointer<Handle>(new Handle(arr));
}

File: greenland/bif.go
/** @copyright 2025 Sean Kasun */
package main

import (
	"bytes"
	"compress/zlib"
	"errors"
	"io"
	"os"
)

type Bif struct {
	data *File
}

func NewBif(filename string) (*Bif, error) {
	data, err := os.ReadFile(filename)
	if err != nil {
		return nil, err
	}

	return &Bif{data: NewFile(data)}, nil
}

const BifMagic = "BIFF"
const BifCompressed = "BIFC"
const BifV1 = "V1  "
const BifV10 = "V1.0"

func (b *Bif) Get(tileset int, index int) (*File, error) {
	b.data.Seek(0)
	key := b.data.R4()
	if key != BifMagic && key != BifCompressed {
		return nil, errors.New("Not a BIF file")
	}
	compressed := key == BifCompressed

	ver := b.data.R4()
	if ver != BifV1 && ver != BifV10 {
		return nil, errors.New("Invalid BIF version")
	}
	if tileset == 0 {
		if compressed {
			h, err := b.getBlock(16, 4)
			if err != nil {
				return nil, err
			}
			fileOfs := h.R32() + index*16
			h, err = b.getBlock(fileOfs, 16)
			if err != nil {
				return nil, err
			}
			h.Skip(4) // locator
			offset := h.R32()
			size := h.R32()
			return b.getBlock(offset, size)
		}
		b.data.Skip(8) // num files and num tiles
		fileOfs := b.data.R32() + index*16
		b.data.Seek(fileOfs)
		b.data.Skip(4) // locator
		offset := b.data.R32()
		size := b.data.R32()
		b.data.Seek(offset)
		return NewFile(b.data.Read(size)), nil
	}
	if compressed {
		h, err := b.getBlock(8, 4)
		if err != nil {
			return nil, err
		}
		numFiles := h.R32()
		h, err = b.getBlock(16, 4)
		if err != nil {
			return nil, err
		}
		fileOfs := h.R32() + numFiles*16 + tileset*20
		h, err = b.getBlock(fileOfs, 20)
		if err != nil {
			return nil, err
		}
		h.Skip(4) // locator
		offset := h.R32()
		numTiles := h.R32()
		size := h.R32()
		return b.getBlock(offset, size*numTiles)
	}
	numFiles := b.data.R32()
	b.data.Skip(4) // num tiles
	fileOfs := b.data.R32() + numFiles*16 + tileset*20
	b.data.Seek(fileOfs)
	b.data.Skip(4) // locator
	offset := b.data.R32()
	numTiles := b.data.R32()
	size := b.data.R32()
	b.data.Seek(offset)
	return NewFile(b.data.Read(size * numTiles)), nil
}

func (b *Bif) getBlock(offset int, length int) (*File, error) {
	b.data.Seek(12)
	cur := 0
	cLen := 0
	uLen := 0
	for cur <= offset && !b.data.EOF() {
		b.data.Skip(cLen)
		if !b.data.EOF() {
			uLen = b.data.R32()
			cLen = b.data.R32()
			cur += uLen
		}
	}

	if b.data.EOF() {
		return nil, errors.New("Block not found")
	}
	delta := offset - (cur - uLen)
	dLen := 0
	arr := make([]byte, length)
	for dLen < length {
		in := bytes.NewBuffer(b.data.Read(cLen))
		reader, err := zlib.NewReader(in)
		if err != nil {
			return nil, err
		}
		unpack, err := io.ReadAll(reader)
		if err != nil {
			return nil, err
		}
		uLen -= delta
		if uLen+dLen > length {
			uLen = length - dLen
		}
		copy(arr[dLen:dLen+uLen], unpack[delta:delta+uLen])
		dLen += uLen
		delta = 0
		if dLen < length {
			uLen = b.data.R32()
			cLen = b.data.R32()
		}
	}
	return NewFile(arr), nil
}
